### ISSUE TITLE
[SR-9133] JUnit output format doesn't work if --parallel not selected

### ISSUE THREAD

**Issue by alblue** [apple/swift-package-manager#4752]
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-9133      |
   |Radar             | None         |
   |Original Reporter | @alblue      |
   |Type              | Bug    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Package Manager    |
   |Labels            | Bug, Screened        |
   |Assignee          | anayini (JIRA)      |
   |Priority          | Medium      |

   

   md5: 7a7f61fc246593e1da18eec9acf288a5

  </details>

**Issue Description:**

The JUnit output report should be printed regardless of the setting of --parallel.

Running:  
swift test --xunit-output=/tmp/results.xml

doesn't generate a file at /tmp/results.xml.

Running:  
swift test --xunit-output=/tmp/results.xml --parallel --num-workers=1

does generate a file at /tmp/results.xml

**Comment by Samasaur1** [comment]
Any news on this?

**Comment by jd-sandk** [comment]
It would be nice if this worked as expected, can something please be done to fix this issue.

**Comment by michalszelagsonos** [comment]
Bump for this issue.  We're about to enter 2024, would be great to get something as foundational as this fixed.  Note, this kind of report is very important for CI and observability integrations.

**Comment by grynspan** [comment]
Tracked internally as rdar://129450002. (I'm pretty sure it's tracked with an older radar number too, but I can't find it right now.)

**Comment by Samasaur1** [comment]
> Also, why is the option labeled xunit? It looks like this format is actually based on junit (from Java), while xunit appears to be something that's part of the .NET ecosystem.

There is an xUnit.net project that is for the .NET ecosystem, but xunit has also been used to refer to a family of testing frameworks (the idea being that the x refers to the language). See [Wikipedia](https://en.wikipedia.org/wiki/XUnit).

**Comment by grynspan** [comment]
The naming of the flag predates me joining the team. For what it's worth, swift-testing does call it JUnit internally. 🤷🏻‍♂️

**Comment by michalszelagsonos** [comment]
Bumping this issue.  We're approaching 2025, can this fix be prioritized?

**Comment by plemarquand** [comment]
Are we sure this is still an issue? I'm using a nightly toolchain and I just created a simple test project with two XCTest suites and ran:

`swift test --xunit-output=./results.xml`:

```
<?xml version="1.0" encoding="UTF-8"?>

<testsuites>
<testsuite name="TestResults" errors="0" tests="2" failures="0" time="0.095531209">
<testcase classname="AAATests.MySuite" name="testExample" time="0.046890917">
</testcase>
<testcase classname="AAATests.MySuite2" name="testExample2" time="0.048640292">
</testcase>
</testsuite>
</testsuites
```

And then: `swift test --xunit-output=./results.xml --parallel --num-workers=1`

```
<?xml version="1.0" encoding="UTF-8"?>

<testsuites>
<testsuite name="TestResults" errors="0" tests="2" failures="0" time="0.09669525100000001">
<testcase classname="AAATests.MySuite" name="testExample" time="0.049329584">
</testcase>
<testcase classname="AAATests.MySuite2" name="testExample2" time="0.047365667">
</testcase>
</testsuite>
</testsuites>
```

**Comment by grynspan** [comment]
I am suspicious of your results as Swift Package Manager doesn't know how to scrape results from XCTest when it runs serially. 🤨

**Comment by plemarquand** [comment]
Yeah looking at the code in SwiftPM I noticed that, but results.xml seems to be being generated by the swiftpm-testing-helper in the default non parallel case, not SwiftPM. You can see yourself by running:

`
DYLD_FRAMEWORK_PATH=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks DYLD_LIBRARY_PATH=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/libexec/swift/pm/swiftpm-testing-helper --test-bundle-path /Users/plemarquand/work/testapps/AAA/.build/arm64-apple-macosx/debug/AAAPackageTests.xctest/Contents/MacOS/AAAPackageTests --xunit-output=./results.xml
`

**Comment by grynspan** [comment]
The helper tool only knows how to invoke Swift Testing. What you're seeing is the XML output from Swift Testing (which _does_ know how to generate it when tests run serially.)

**Comment by plemarquand** [comment]
Sigh, I had a typo in my command to remove old .xml, I am seeing the behaviour described in the initial report now.

**Comment by plemarquand** [comment]
Seems like this could be implemented with https://github.com/swiftlang/swift-package-manager/pull/6861

**Comment by grynspan** [comment]
IIRC there were a number of problems with that PR but it's been _ages_ since I looked at it.

And yes, the `XCTestObservation` protocol is only implemented in XCTest.framework, not corelibs-xctest. That doesn't mean it _can't_ be implemented (approximately, ignoring ObjC minutiae) but it's always a question of engineering resources.

**Comment by grynspan** [comment]
Note that `XCTestObservation` has a fair bit of conceptual overlap with the JSON stream in Swift Testing and, I think ideally, we'd want to move XCTest toward Swift Testing rather than reimplementing yet more features in corelibs-xctest. (But that's a question for the upcoming Testing Workgroup Whatchamacallit.)
